napari/components/dims.py

```diff
@@ -67,285 +66,268 @@ class Dims(EventedModel):
         ``displayed`` dimensions.
     """
 
-    # fields
+    # fields (order matters for validation! Previous fields are available for following fields)
     ndim: int = 2
     ndisplay: Literal[2, 3] = 2
+    axis_labels: EventedList[str] = ['0', '1']
+    range: NestableEventedList[NestableEventedList[float]] = [[0, 2], [0, 2]]
+    span: NestableEventedList[NestableEventedList[float]] = [[0, 0], [0, 0]]
+    step: EventedList[float] = [1, 1]
+    order: EventedList[int] = [0, 1]
     last_used: int = 0
-    range: Tuple[Tuple[float, float, float], ...] = ()
-    current_step: Tuple[int, ...] = ()
-    order: Tuple[int, ...] = ()
-    axis_labels: Tuple[str, ...] = ()
-
     # private vars
     _scroll_progress: int = 0
 
-    # validators
-    @validator('axis_labels', pre=True)
-    def _string_to_list(v):
-        if isinstance(v, str):
-            return list(v)
-        return v
-
-    @root_validator
-    def _check_dims(cls, values):
-        """Check the consitency of dimensionaity for all attributes
-
-        Parameters
-        ----------
-        values : dict
-            Values dictionary to update dims model with.
-        """
+    class Config:
+        computed_fields = {
+            '_span_step': ['span', 'range', 'step'],
+            'nsteps': ['range', 'step'],
+            'thickness': ['span'],
+            '_thickness_step': ['span', 'step'],
+            'point': ['span'],
+            '_point_step': ['span', 'range', 'step'],
+            'current_step': ['span', 'range', 'step'],
+            'displayed': ['order', 'ndisplay'],
+            'not_displayed': ['order', 'ndisplay'],
+            'displayed_order': ['order', 'ndisplay'],
+        }
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # to be removed if/when deprecating current_step
+        self.events.span.connect(self.events.current_step)
+
+    @root_validator(skip_on_failure=True)
+    def _enforce_ndim(cls, values):
+        # This validator should not be split, or setting attributes will only trigger
+        # invididual validators, potentially not updating the other fields accordingly
```

Comment:
This is an annoyign thing with `Dims`: on one hand, we want to be able to simply do `Dims.ndim = 3` and get a usable dims; on the other hand, in some cases we would like to simply complain if someone does `Dims.range = too_long_list`. Currently, the behavior is a broken hybrid, due to issues with parent validation (see later comments).